File: android/app/src/main/java/com/videoconverter/android/lan/LanMedia.kt
package com.videoconverter.android.lan

enum class LanPreviewKind { Video, Audio, Pdf, Image, File }

fun lanPreviewKind(fileName: String): LanPreviewKind = when (fileName.substringAfterLast('.', "").lowercase()) {
    "mp4", "mov", "mkv", "webm", "avi" -> LanPreviewKind.Video
    "mp3", "m4a", "wav", "ogg", "flac", "amr" -> LanPreviewKind.Audio
    "pdf" -> LanPreviewKind.Pdf
    "jpg", "jpeg", "png", "webp", "gif", "bmp" -> LanPreviewKind.Image
    else -> LanPreviewKind.File
}

fun lanContentType(fileName: String): String = when (fileName.substringAfterLast('.', "").lowercase()) {
    "mp4" -> "video/mp4"
    "mov" -> "video/quicktime"
    "mkv" -> "video/x-matroska"
    "webm" -> "video/webm"
    "avi" -> "video/x-msvideo"
    "mp3" -> "audio/mpeg"
    "m4a" -> "audio/mp4"
    "wav" -> "audio/wav"
    "ogg" -> "audio/ogg"
    "flac" -> "audio/flac"
    "amr" -> "audio/amr"
    "jpg", "jpeg" -> "image/jpeg"
    "png" -> "image/png"
    "webp" -> "image/webp"
    "gif" -> "image/gif"
    "bmp" -> "image/bmp"
    "pdf" -> "application/pdf"
    "txt" -> "text/plain; charset=utf-8"
    else -> "application/octet-stream"
}

fun lanContentDisposition(fileName: String, inline: Boolean = false): String {
    val safe = fileName.replace(Regex("[\r\n\"]"), "_")
    val encoded = java.net.URLEncoder.encode(safe, "UTF-8").replace("+", "%20")
    val kind = if (inline) "inline" else "attachment"
    return "$kind; filename=\"$safe\"; filename*=UTF-8''$encoded"
}

fun isLanContentLocation(location: String): Boolean =
    location.startsWith("content:", ignoreCase = true)

sealed class LanByteRange {
    data object Whole : LanByteRange()
    data class Partial(val start: Long, val endInclusive: Long) : LanByteRange()
    data object Unsatisfiable : LanByteRange()
}

fun parseLanByteRange(header: String?, total: Long): LanByteRange {
    val raw = header?.trim().orEmpty()
    if (raw.isEmpty()) return LanByteRange.Whole
    if (total <= 0L) return LanByteRange.Unsatisfiable
    if (',' in raw) return LanByteRange.Unsatisfiable
    val prefix = "bytes="
    if (!raw.startsWith(prefix, ignoreCase = true)) return LanByteRange.Unsatisfiable
    val spec = raw.substring(prefix.length)
    val dash = spec.indexOf('-')
    if (dash < 0) return LanByteRange.Unsatisfiable
    val startText = spec.substring(0, dash)
    val endText = spec.substring(dash + 1)
    val start = startText.toLongOrNull() ?: return LanByteRange.Unsatisfiable
    if (start < 0L || start >= total) return LanByteRange.Unsatisfiable
    val end = if (endText.isEmpty()) total - 1 else endText.toLongOrNull() ?: return LanByteRange.Unsatisfiable
    if (end < start) return LanByteRange.Unsatisfiable
    return LanByteRange.Partial(start, minOf(end, total - 1))
}

fun lanContentRangeValue(start: Long, endInclusive: Long, total: Long): String =
    "bytes $start-$endInclusive/$total"

fun lanUnsatisfiableContentRange(total: Long): String = "bytes */$total"

fun copyLanRange(input: java.io.InputStream, output: java.io.OutputStream, start: Long, length: Long) {
    var skipped = 0L
    while (skipped < start) {
        val n = input.skip(start - skipped)
        if (n <= 0L) {
            if (input.read() < 0) return
            skipped += 1
        } else {
            skipped += n
        }
    }
    var remaining = length
    val buffer = ByteArray(8192)
    while (remaining > 0) {
        val want = minOf(buffer.size.toLong(), remaining).toInt()
        val read = input.read(buffer, 0, want)
        if (read < 0) return
        output.write(buffer, 0, read)
        remaining -= read
    }
}

File: ios/LiteTrans/Domain/LanMedia.swift
import Foundation

public enum LanPreviewKind: Equatable, Sendable {
    case video, audio, pdf, image, file
}

public func lanPreviewKind(_ fileName: String) -> LanPreviewKind {
    switch lanFileExtension(fileName) {
    case "mp4", "mov", "mkv", "webm", "avi": return .video
    case "mp3", "m4a", "wav", "ogg", "flac", "amr": return .audio
    case "pdf": return .pdf
    case "jpg", "jpeg", "png", "webp", "gif", "bmp": return .image
    default: return .file
    }
}

public func lanContentType(_ fileName: String) -> String {
    switch lanFileExtension(fileName) {
    case "mp4": return "video/mp4"
    case "mov": return "video/quicktime"
    case "mkv": return "video/x-matroska"
    case "webm": return "video/webm"
    case "avi": return "video/x-msvideo"
    case "mp3": return "audio/mpeg"
    case "m4a": return "audio/mp4"
    case "wav": return "audio/wav"
    case "ogg": return "audio/ogg"
    case "flac": return "audio/flac"
    case "amr": return "audio/amr"
    case "jpg", "jpeg": return "image/jpeg"
    case "png": return "image/png"
    case "webp": return "image/webp"
    case "gif": return "image/gif"
    case "bmp": return "image/bmp"
    case "pdf": return "application/pdf"
    case "txt": return "text/plain; charset=utf-8"
    default: return "application/octet-stream"
    }
}

public func lanContentDisposition(_ fileName: String, inline: Bool = false) -> String {
    let safe = fileName.replacingOccurrences(of: "[\r\n\"]", with: "_", options: .regularExpression)
    let encoded = lanQueryEncode(safe).replacingOccurrences(of: "+", with: "%20")
    let kind = inline ? "inline" : "attachment"
    return "\(kind); filename=\"\(safe)\"; filename*=UTF-8''\(encoded)"
}

public func isLanContentLocation(_ location: String) -> Bool {
    location.lowercased().hasPrefix("content:")
}

func lanFileExtension(_ fileName: String) -> String {
    guard let dot = fileName.lastIndex(of: "."), dot != fileName.endIndex else { return "" }
    return String(fileName[fileName.index(after: dot)...]).lowercased()
}

public enum LanByteRange: Equatable, Sendable {
    case whole
    case partial(start: Int64, endInclusive: Int64)
    case unsatisfiable
}

public func parseLanByteRange(_ header: String?, total: Int64) -> LanByteRange {
    let raw = header?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    if raw.isEmpty { return .whole }
    if total <= 0 { return .unsatisfiable }
    if raw.contains(",") { return .unsatisfiable }
    let prefix = "bytes="
    guard raw.lowercased().hasPrefix(prefix) else { return .unsatisfiable }
    let spec = String(raw.dropFirst(prefix.count))
    guard let dash = spec.firstIndex(of: "-") else { return .unsatisfiable }
    let startText = String(spec[..<dash])
    let endText = String(spec[spec.index(after: dash)...])
    guard let start = Int64(startText) else { return .unsatisfiable }
    if start < 0 || start >= total { return .unsatisfiable }
    let end: Int64
    if endText.isEmpty {
        end = total - 1
    } else if let parsed = Int64(endText) {
        end = parsed
    } else {
        return .unsatisfiable
    }
    if end < start { return .unsatisfiable }
    return .partial(start: start, endInclusive: min(end, total - 1))
}

public func lanContentRangeValue(start: Int64, endInclusive: Int64, total: Int64) -> String {
    "bytes \(start)-\(endInclusive)/\(total)"
}

public func lanUnsatisfiableContentRange(_ total: Int64) -> String {
    "bytes */\(total)"
}

public func lanReadyFileResponse(_ sized: LanHttpResponse, opened: Bool) -> LanHttpResponse {
    if sized.filePath == nil || opened { return sized }
    return LanHttpResponse(
        status: 404,
        contentType: "text/plain; charset=utf-8",
        body: Data("Not Found".utf8),
        sendBody: sized.sendBody
    )
}

public func applyLanResponseRange(_ response: LanHttpResponse, total: Int64) -> LanHttpResponse {
    guard response.filePath != nil else { return response }
    var next = response
    switch parseLanByteRange(response.rangeHeader, total: total) {
    case .whole:
        next.status = 200
        next.headers["Content-Length"] = String(total)
        next.byteStart = 0
        next.byteLength = nil
        return next
    case let .partial(start, endInclusive):
        let length = endInclusive - start + 1
        next.status = 206
        next.headers["Content-Range"] = lanContentRangeValue(start: start, endInclusive: endInclusive, total: total)
        next.headers["Content-Length"] = String(length)
        next.byteStart = start
        next.byteLength = length
        return next
    case .unsatisfiable:
        next.status = 416
        next.headers["Content-Range"] = lanUnsatisfiableContentRange(total)
        next.body = Data()
        next.filePath = nil
        next.sendBody = false
        return next
    }
}

public func copyLanRange(from data: Data, start: Int64, length: Int64) -> Data {
    let begin = Int(start)
    let end = min(data.count, begin + Int(length))
    guard begin >= 0, begin < data.count, end > begin else { return Data() }
    return data.subdata(in: begin..<end)
}
